[autoscaler] Fix update/terminate race condition

DmitriGekhtman opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

It seems it's possible for a node to be terminated while a NodeUpdaterThread is trying to update it.
This causes the update to fail. The autoscaler then repeatedly tries to terminate the node again which leads to an error. The autoscaler continues to try to delete the non-existent node until the autoscaler exceeds `max_failures` and stops.

This PR
(1) Modifies autoscaler logic to avoid the race condition
(2) Modifies `kubernetes_node_provider.terminate_node()` so that it handles the 404 error if it tries to delete a missing pod. 

See https://github.com/ray-project/ray/issues/14996 .

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

Closes https://github.com/ray-project/ray/issues/14996

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

Open to any testing suggestions / any ideas on how to replicate the error in https://github.com/ray-project/ray/issues/14996 .

wuisawesome commented:
Is this a race condition?

If so, shouldn't we fix it with a trying to terminate all the nodes and catching any exceptions (ideally terminating a node twice shouldn't be an error right?). 

Also, wouldn't we need to do this in all the other places that we terminate nodes?

DmitriGekhtman commented:
> Is this a race condition?
> 
> If so, shouldn't we fix it with a trying to terminate all the nodes and catching any exceptions (ideally terminating a node twice shouldn't be an error right?).
> 
> Also, wouldn't we need to do this in all the other places that we terminate nodes?

Different node providers produce different errors when you try to terminate a node that's not there, so like you said we would have to catch any exception each time we call `provider.terminate_nodes` ---
```
try:
    provider.terminate_nodes
except Exception as e:
    blah
```
which I think is not so great -- would ideally want to catch a specific exception.

DmitriGekhtman commented:
Some additional error handling could make things safer, though -- in addition to the change in this PR, we could replace each autoscaler call to `self.provider.non_terminated_nodes` to a call to a method like this.

```
def _terminate_nodes(self, node_ids):
    try:
        self.provider.terminate_nodes(node_ids)
    except Exception:
        logger.warning("Failed to terminate nodes. Trying again.")
        active_nodes = self.workers()
        for node_id in node_ids:
            if node_id in active_nodes:
                self.provider.terminate_node(node_id)
            else:
                logger.warning("Node {} has already been terminated.".format(node_id))
```
and something similar for calls to `self.provider.terminate_node`.

Idea is that node termination would be safer, but also there's a warning logged somewhere that you've tried to delete a non-existent node, which is not a thing that should happen. 

@wuisawesome What do you think?

DmitriGekhtman commented:
> @DmitriGekhtman This function (`def _terminate_nodes(self, node_ids):`) looks reasonable.
> 
> My concern is that this does not close #14996 ([as per here](https://github.com/ray-project/ray/issues/14996#issuecomment-810538992)).

Good point! 
Will wrap the body of `_get_node_type` with
```
try:
except Exception:
    return "unknown"
```

DmitriGekhtman commented:
> > @DmitriGekhtman This function (`def _terminate_nodes(self, node_ids):`) looks reasonable.
> > My concern is that this does not close #14996 ([as per here](https://github.com/ray-project/ray/issues/14996#issuecomment-810538992)).
> 
> Good point!
> Will wrap the body of `_get_node_type` with
> 
> ```
> try:
> except Exception:
>     return "unknown"
> ```

Hmm, also will have these messages printed only for the "active_nodes_to_terminate".

DmitriGekhtman commented:
@AmeerHajAli Oh oops -- I now understand your question about max_workers. I added the new logic under the wrong `if nodes_to_terminate:`.

DmitriGekhtman commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,25 @@ def _update(self):
-                del self.updaters[node_id]
-            if nodes_to_terminate:
-                self.provider.terminate_nodes(nodes_to_terminate)
+                self.provider.terminate_nodes(active_nodes_to_terminate)
```
The diff is hard to read -- I moved the `logger.error` / `self.event_summarizer` part into a for loop under the `if nodes_to_terminate` block.

ijrsvt commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,25 @@ def _update(self):
+                active_nodes = self.workers()
+                active_nodes_to_terminate = [
+                    node for node in nodes_to_terminate if node in active_nodes
+                ]
+                for node_id in active_nodes_to_terminate:
```
Should we debug log/info log on nodes that are in `nodes_to_terminate` and not in `active_nodes`?

DmitriGekhtman replied:
Good idea.
That also makes things a bit more symmetric.

DmitriGekhtman commented:
Will add unit tests for the `terminate_workers` function and the post-update processing logic.

AmeerHajAli reviewed (dismissed):
Instead of doing that I suggest 2 solutions:
1) make terminating a terminated node OK (I don't see why that would be a problem).
2) take a look at my suggestion (lets start with this solution)

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
-                self.provider.terminate_nodes(nodes_to_terminate)
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
```
```suggestion
                non_terminated_nodes = self.workers()
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
                     nodes_to_terminate.append(node_id)
                     self.num_failed_updates[node_id] += 1
                 del self.updaters[node_id]
             if nodes_to_terminate:
```
```suggestion
            if failed_nodes:
```

AmeerHajAli replied:
```suggestion
            if failed_nodes:
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
+                active_nodes_to_terminate = []
```
```suggestion
                nodes_to_terminate = []
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
+                active_nodes_to_terminate = []
+                for node_id in nodes_to_terminate:
```
```suggestion
                for node_id in failed_nodes:
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
+                active_nodes = self.workers()
+                active_nodes_to_terminate = []
+                for node_id in nodes_to_terminate:
+                    if node_id in active_nodes:
```
```suggestion
                    if node_id in non_terminated_nodes:
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
+                                     " Node already terminated.")
+                self.terminate_workers(active_nodes_to_terminate)
 
             nodes = self.workers()
```
```suggestion
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,18 +267,30 @@ def _update(self):
                 del self.updaters[node_id]
             if nodes_to_terminate:
-                self.provider.terminate_nodes(nodes_to_terminate)
+                # Some nodes in nodes_to_terminate may have been terminated
+                # during an update. Only terminate currently active nodes.
+                active_nodes = self.workers()
+                active_nodes_to_terminate = []
+                for node_id in nodes_to_terminate:
+                    if node_id in active_nodes:
+                        active_nodes_to_terminate.append(node_id)
+                        logger.error(f"StandardAutoscaler: {node_id}:"
+                                     " Terminating. Failed to setup/initialize"
+                                     " node.")
+                        self.event_summarizer.add(
+                            "Removing {} nodes of type " +
+                            self._get_node_type(node_id) + " (launch failed).",
+                            quantity=1,
+                            aggregate=operator.add)
+                    else:
+                        logger.error(f"StandardAutoscaler: {node_id}: "
+                                     " Failed to setup/initialize node."
+                                     " Node already terminated.")
+                self.terminate_workers(active_nodes_to_terminate)
```
```suggestion
                # Some nodes in failed_nodes may have been terminated
                # during an update (for being idle after missing a heartbeat). 
                # Only terminate currently non terminated nodes.
                non_terminated_nodes = self.workers()
                nodes_to_terminate = []
                for node_id in failed_nodes:
                    if node_id in non_terminated_nodes:
                        nodes_to_terminate.append(node_id)
                        logger.error(f"StandardAutoscaler: {node_id}:"
                                     " Terminating. Failed to setup/initialize"
                                     " node.")
                        self.event_summarizer.add(
                            "Removing {} nodes of type " +
                            self._get_node_type(node_id) + " (launch failed).",
                            quantity=1,
                            aggregate=operator.add)
                if nodes_to_terminate:
                    self.provider.terminate_nodes(nodes_to_terminate)
                    for node in nodes_to_terminate:
                        self.node_tracker.untrack(node)
                    nodes = self.workers()
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -309,6 +321,27 @@ def _update(self):
         logger.info(self.info_string())
         legacy_log_info_string(self, nodes)
 
+    def terminate_workers(self, node_ids: List[str]) -> None:
+        """
+        Terminate workers with the supplied node_ids.
+
+        Does some exception handling to avoid failure in case the autoscaler
+        tries to terminate the same node multiple times.
+        """
+        try:
+            self.provider.terminate_nodes(node_ids)
+        except Exception:
+            logger.error("Failed to terminate nodes. Trying again.")
+            active_nodes = self.workers()
+            for node_id in node_ids:
+                if node_id in active_nodes:
+                    logger.error(f"StandardAutoscaler: {node_id}: "
+                                 " Trying again to terminate.")
+                    self.provider.terminate_node(node_id)
+                else:
+                    logger.error(f"StandardAutoscaler: {node_id}: "
+                                 " Node already terminated.")
+
```
```suggestion
```

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -598,10 +631,13 @@ def recover_if_needed(self, node_id, now):
         self.updaters[node_id] = updater
 
     def _get_node_type(self, node_id: str) -> str:
```
Keep the function the same

AmeerHajAli commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -731,7 +767,7 @@ def kill_workers(self):
         nodes = self.workers()
         if nodes:
-            self.provider.terminate_nodes(nodes)
+            self.terminate_workers(nodes)
```
```suggestion
```

DmitriGekhtman commented:
> 1. make terminating a terminated node OK (I don't see why that would be a problem).

This would in any case have to be done with exception handling in the autoscaler, otherwise we would have to guarantee that none of the NodeProviders raise exceptions in this situation and add a requirement to the NodeProviderInterface that terminating the same node twice has to be handled there.

AmeerHajAli commented:
@DmitriGekhtman , Done. Can you actually move everything under `if completed` to a new function `handle_completed_updaters` ?

DmitriGekhtman commented:
@AmeerHajAli do you think we should do additional error handling in the autoscaler for nodes that are terminated twice?

The PR in its current state resolves one race condition leading to double termination.
This PR fixed another race condition leading to the same behavior:
https://github.com/ray-project/ray/pull/14976

I imagine problems of this sort could arise again, and it could help to be defensive.

ijrsvt commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,20 +268,32 @@ def _update(self):
+                        logger.error(f"StandardAutoscaler: {node_id}:"
+                                     " Terminating. Failed to setup/initialize"
+                                     " node.")
+                        self.event_summarizer.add(
+                            "Removing {} nodes of type " +
+                            self._get_node_type(node_id) + " (launch failed).",
+                            quantity=1,
+                            aggregate=operator.add)
```
Can we add an `else` statement here that debug/info logs double terminations?

DmitriGekhtman replied:
Add an else with a warning.

ijrsvt commented on python/ray/autoscaler/_private/autoscaler.py:
```diff
@@ -267,20 +268,32 @@ def _update(self):
+                if nodes_to_terminate:
+                    self.provider.terminate_nodes(nodes_to_terminate)
+                    for node in nodes_to_terminate:
+                        self.node_tracker.untrack(node)
```
Was this missing in the previous implementation?

DmitriGekhtman replied:
I don't think so.

We do need to also un-track nodes which failed the update and were already terminated.

ijrsvt commented:
> I imagine problems of this sort could arise again, and it could help to be defensive.

+1 on being defensive here. That being said, I think the current behavior (only terminating nodes that are in `nodes` is okay).

ijrsvt reviewed (approved):
This looks good to me at this point!

DmitriGekhtman commented:
>  Can you actually move everything under `if completed` to a new function `handle_completed_updaters` ?

@AmeerHajAli 
I think it might actually be better to do this refactor in another PR.
Also -- it's not immediately obvious how to do this cleanly, in particular how to update the variable `nodes`.
But for sure we should separate each step of `_update` that creates or deletes nodes into a separate function.

Did we open an issue a while ago to clean up autoscaler code?

DmitriGekhtman commented:
Will do these things:
- Attempt a unit test for the new autoscaler logic
- Add a tiny bit of logic to k8s operator test to check the error handling on terminate_node

AmeerHajAli commented:
I think it might be good sometimes to fail to see what kind of issues could happen. If we add all this logic and it ends up not ever happening, it becomes dead code.

DmitriGekhtman commented:
> I think it might be good sometimes to fail to see what kind of issues could happen. If we add all this logic and it ends up not ever happening, it becomes dead code.

Yeah, I agree. There is some sort of balance to be achieved with error handling -- keep the system running reliably but don't swallow internal inconsistencies.

DmitriGekhtman commented:
Added a unit test to validate the new completed-update-processing logic.
Unfortunately, it was interacting badly with the other tests in `test_autoscaler.py` so I had to make a new file `test_autoscaler2.py` for the new unit test.

Also added a test of the new K8s node provider exception handling.

DmitriGekhtman commented:
ugh, new test is flaky

AmeerHajAli commented:
> ugh, new test is flaky

What is going on?

DmitriGekhtman commented:
Test should work now.

DmitriGekhtman commented:
Test failures are unrelated, will rebase and run once more just for good measure.

DmitriGekhtman commented:
good to merge

DmitriGekhtman commented:
windows test_scheduling is known flaky.
@AmeerHajAli or @ijrsvt , could you merge this?
